octree_web_viewer/client/main.ts

```diff
@@ -20,124 +20,203 @@ import { FirstPersonController } from './control';
-      .step(1)
-      .onChange(() => {
+    private camera: THREE.PerspectiveCamera;
+    private scene: THREE.Scene;
+    private controller: FirstPersonController;
+    private viewer: OctreeViewer;
+    private renderer: THREE.WebGLRenderer;
+    private lastFrustumUpdateTime: number;
+    private lastMoveTime: number;
+    private needsRender: boolean;
+    private octreeIdControl: dat.GUIController;
+    private gui: dat.GUI;
+    private guiRenderControls: dat.GUI;
+    public octreeId: string;  // octree identifier
+
+    private fetchDefaultOctreeId(): Promise<string> {
+        const request = new Request(
+            `/init_tree`,
+            {
+                method: 'GET',
+                credentials: 'same-origin',
+            }
+        );
+
+        const result = window
+            .fetch(request)
+            .then((response) => response.text()) // todo error handling?
+            .then((body) => {
+                return body;
+            });
+        return result;
+    }
+
+    private initOctreeViewer(octreeId: string) {
+        this.viewer = new OctreeViewer(this.scene, () => {
+            this.needsRender = true;
+        }, octreeId);
+    }
+
+    private addControls() {
+        this.guiRenderControls = this.gui.addFolder('Render controls');
+        this.guiRenderControls
+            .add(this.viewer.material.uniforms['size'], 'value')
+            .name('Point size')
+            .onChange(() => {
+                this.needsRender = true;
+            });
+        this.guiRenderControls
+            .add(this.viewer.material.uniforms['alpha'], 'value', 0, 1)
+            .name('Transparency')
+            .onChange(() => {
+                this.viewer.alphaChanged();
+                this.needsRender = true;
+            });
+        this.guiRenderControls
+            .add(this.viewer.material.uniforms['gamma'], 'value')
+            .name('Gamma')
+            .onChange(() => {
+                this.needsRender = true;
+            });
+        this.guiRenderControls
+            .add(this.viewer, 'maxLevelToDisplay', 0, 7)
+            .name('Moving details')
+            .step(1)
+            .onChange(() => {
+                this.needsRender = true;
+            });
+    }
+
+    private initCamera() {
+        const renderArea = document.getElementById('renderArea');
+        const VIEW_ANGLE = 45;
+        const NEAR = 0.1;
+        const FAR = 10000;
+        this.camera = new THREE.PerspectiveCamera(
+            VIEW_ANGLE,
+            renderArea.clientWidth / renderArea.clientHeight,
+            NEAR,
+            FAR
+        );
+        this.camera.position.z = 150;
+        this.camera.updateMatrix();
+        this.camera.updateMatrixWorld(false);
+    }
+
+    private initScene() {
+        this.scene = new THREE.Scene();
+        this.scene.add(this.camera);
+    }
+
+    private initRenderer() {
+        const renderArea = document.getElementById('renderArea');
```

Comment:
Please create a method `getViewportSize` returning `{width, height}` and reuse it when creating renderer and camera. Also you can make DOM element you are rendering to a class property. This way you'll get rid of those repeated `document.getElementById('renderArea')` calls.